keps/sig-api-machinery/2876-crd-validation-expression-language/README.md

```diff
@@ -260,6 +260,23 @@ Example Validation Rules:
 
 - Each validation rule has an optional 'message' field for the error message that
 will be surfaced when the validation rule evaluates to false.
+- As an alternative to the `message` field, there is also a
+`messageExpression` field that represents a CEL expression that evaluates to a
+string that will be used when the validation rule evaluates to false. If the
+`messageExpression` field does not evaluate to a string, then the CRD write will fail. If the
```

Comment:
I guess user will not care about the messageExpression failures as much as validation expression failures. Do we really want to fail them if messageExpression fail? Maybe consider to log the failure somewhere instead of fail the request?

Reply:
I think in practice a lot of type errors with `messageExpression` would look like:

```
"someVar can be 6 at most, but is: " + someVar
```

And `someVar` is an integer or double, or some other non-string type. This can be fixed by using `string(someVar)` instead, and seems to me to be the sort of thing users would want to know about at CRD write time.

If we want to opt for complete consistency with CEL for admission policies, then falling back would be a good option here. Still, I'm in favor of failing the write if `messageExpression` isn't a string - it seems we can easily eliminate a whole class of errors.

Reply:
We should absolutely type check `messageExpression` statically (When CRDs are created/updated). Checking that the expression evaluates to string seems right too. I would prefer keep invalid `messageExpression`  out of CRDs entirely, to the extent we can detect it.

Reply:
+1 to check `messageExpression` when CRD create/update. Question is about how we treat the failure(if worth to fail the request because of `messageExpression` failure). I guess it is ok to log it or have a metrics for it but maybe let the request pass with invalid `messageExpression` eliminated?

Reply:
> if worth to fail the request because of messageExpression failure

If we're evaluating `messageExpression` that means the request has already failed.

Reply:
(marking as resolved since now `messageExpression` is discarded if it doesn't statically eval to a string)

Reply:
Why not just fail CRD validation if `messageExpression` is not of type string?